crates/arrow_tools/src/record_batch.rs

```diff
@@ -0,0 +1,267 @@
+    ParseTimestampFromString { source: chrono::ParseError },
+
+    #[snafu(display("Unable to convert {} array from {:?} to {:?}", name, from, to))]
+    ConvertArrowArray {
+        from: DataType,
+        to: DataType,
+        name: String,
+    },
+
+    #[snafu(display("Unable to try from int: {}", source))]
+    ConvertInt { source: TryFromIntError },
+
+    #[snafu(display("Unable to use null for field {field}"))]
+    UnableToUseNullForField { field: String },
+
+    #[snafu(display("Unable to downcast array {field} to {downcast_to}"))]
+    UnableToDowncast { field: String, downcast_to: String },
+}
+
+/// Cast a given record batch into a new record batch with the given schema.
+///
+/// # Errors
+///
+/// This function will return an error if the record batch cannot be casted.
+#[allow(clippy::needless_pass_by_value)]
+pub fn try_cast_to(record_batch: RecordBatch, schema: SchemaRef) -> Result<RecordBatch> {
+    let existing_schema = record_batch.schema();
+
+    if schema.contains(&existing_schema) {
+        return record_batch
+            .with_schema(schema)
+            .context(ConvertRecordBatchSnafu);
+    }
+
+    let cols = schema
+        .fields()
+        .into_iter()
+        .map(|field| {
+            if let (Ok(existing_field), Some(column)) = (
+                record_batch.schema().field_with_name(field.name()),
+                record_batch.column_by_name(field.name()),
+            ) {
+                if field.contains(existing_field) {
+                    Ok(Arc::clone(column))
+                } else {
+                    try_cast_array(Arc::clone(column), existing_field, field)
+                }
+            } else if field.is_nullable() {
+                Ok(new_null_array(field.data_type(), record_batch.num_rows()))
+            } else {
+                UnableToUseNullForFieldSnafu {
+                    field: field.name(),
+                }
+                .fail()
+            }
+        })
+        .collect::<Result<Vec<Arc<dyn Array>>>>()?;
+
+    RecordBatch::try_new(schema, cols).context(ConvertRecordBatchSnafu)
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn try_cast_array(
+    array: Arc<dyn Array>,
+    existing_field: &Field,
+    field: &Field,
+) -> Result<Arc<dyn Array>> {
+    let result: Arc<dyn Array> = match (existing_field.data_type(), field.data_type()) {
+        (DataType::Int32, DataType::Int64) => {
+            if let Some(array) = array.as_any().downcast_ref::<Int32Array>() {
+                Arc::new(Int64Array::from(
+                    array
+                        .into_iter()
+                        .map(|f| f.map(i64::from))
+                        .collect::<Vec<Option<i64>>>(),
+                ))
+            } else {
+                UnableToDowncastSnafu {
+                    field: field.name(),
+                    downcast_to: "Int32Array".to_string(),
+                }
+                .fail()?
+            }
```

Comment:
```suggestion
            let array = array.as_any().downcast_ref::<Int32Array>().context(
                  UnableToDowncastSnafu {
                      field: field.name(),
                      downcast_to: "Int32Array".to_string(),
                  },
              )?;
              Arc::new(Int64Array::from(
                  array
                      .into_iter()
                      .map(|f| f.map(i64::from))
                      .collect::<Vec<Option<i64>>>(),
              ))
```

I find this a bit nicer.

Reply:
good to know.